### PR TITLE
Basic operator semantics

### LINKED ISSUE #25: [FEATURE REQUEST] Semantics for basic operations
**Description**
Currently, the semantics for some basic operations like let-assignment, bitwise operation and arithmetic operation where missing.

**Motivation**
We can add some semantics for basic operations.

### PR DESCRIPTION
**Pull request by simonegazza**
The aim of this pull request is to add basic semantics for some operators like arithmetic ones, bitwise and let-assignement

Closes #25.

### CODE REVIEW

File: rust-lisa/src/main/java/it/unipr/cfg/expression/comparison/RustAndExpression.java [hunk 1]
```diff
@@ -14,6 +14,8 @@
 import it.unive.lisa.program.cfg.statement.BinaryExpression;
```
Comment (VincenzoArceri) [inline rust-lisa/src/main/java/it/unipr/cfg/expression/comparison/RustAndExpression.java:14]: Does Rust implement a short-circut/lazy semantics? If so, the and semantics should be revised.

File: rust-lisa/src/main/java/it/unipr/cfg/expression/comparison/RustOrExpression.java [hunk 2]
```diff
@@ -14,6 +14,8 @@
 import it.unive.lisa.program.cfg.statement.BinaryExpression;
```
Comment (VincenzoArceri) [inline rust-lisa/src/main/java/it/unipr/cfg/expression/comparison/RustOrExpression.java:14]: Does Rust implement a short-circut/lazy semantics? If so, the or semantics should be revised.

File: rust-lisa/src/main/java/it/unipr/cfg/expression/numeric/RustDivExpression.java [hunk 3]
```diff
@@ -46,8 +48,17 @@ T extends TypeDomain<T>> AnalysisState<A, H, V, T> binarySemantics(
...
 					SymbolicExpression left, SymbolicExpression right, StatementStore<A, H, V, T> expressions)
 					throws SemanticException {
-		// TODO too coarse
-		return state.top();
+
+		AnalysisState<A, H, V, T> result = state.bottom();
+
+		for (Type leftType : left.getRuntimeTypes())
+			for (Type rightType : right.getRuntimeTypes())
+				if (leftType.canBeAssignedTo(rightType) && rightType.canBeAssignedTo(leftType))
```
Comment (VincenzoArceri) [inline rust-lisa/src/main/java/it/unipr/cfg/expression/numeric/RustDivExpression.java:56]: We should also check that left and right types are numeric types.

File: rust-lisa/src/main/java/it/unipr/cfg/expression/numeric/RustModExpression.java [hunk 4]
```diff
@@ -46,8 +48,17 @@ T extends TypeDomain<T>> AnalysisState<A, H, V, T> binarySemantics(
...
 					SymbolicExpression left, SymbolicExpression right, StatementStore<A, H, V, T> expressions)
 					throws SemanticException {
-		// TODO too coarse
-		return state.top();
+
+		AnalysisState<A, H, V, T> result = state.bottom();
+
+		for (Type leftType : left.getRuntimeTypes())
+			for (Type rightType : right.getRuntimeTypes())
+				if (leftType.canBeAssignedTo(rightType) && rightType.canBeAssignedTo(leftType))
```
Comment (VincenzoArceri) [inline rust-lisa/src/main/java/it/unipr/cfg/expression/numeric/RustModExpression.java:56]: We should also check that left and right types are numeric. For instance, if left and right types are strings, the guard at line 56 holds and reaches lines 57-58 (same for other numerical expressions).

**Review by VincenzoArceri — CHANGES_REQUESTED** [review]

File: rust-lisa/src/main/java/it/unipr/cfg/expression/comparison/RustOrExpression.java (hunk 2 above)
Comment (simonegazza) [reply rust-lisa/src/main/java/it/unipr/cfg/expression/comparison/RustOrExpression.java:15]: An implementation was added in commit 3a08c7d. Currently we are unable to testi it.

File: rust-lisa/src/main/java/it/unipr/cfg/expression/comparison/RustAndExpression.java (hunk 1 above)
Comment (simonegazza) [reply rust-lisa/src/main/java/it/unipr/cfg/expression/comparison/RustAndExpression.java:15]: An implementation was added in commit 3a08c7d. Currently we are unable to testi it.

File: rust-lisa/src/main/java/it/unipr/cfg/expression/numeric/RustDivExpression.java (hunk 3 above)
Comment (simonegazza) [reply rust-lisa/src/main/java/it/unipr/cfg/expression/numeric/RustDivExpression.java:56]: It is possibile to have operator overload in rust, so it is not guaranteed that those types will be numerical. As agreed in a previous conversation, we should revise those function when we will be modelling operator overload.

File: rust-lisa/src/main/java/it/unipr/cfg/expression/numeric/RustModExpression.java (hunk 4 above)
Comment (simonegazza) [reply rust-lisa/src/main/java/it/unipr/cfg/expression/numeric/RustModExpression.java:56]: It is possibile to have operator overload in rust, so it is not guaranteed that those types will be numerical. As agreed in a previous conversation, we should revise those function when we will be modelling operator overload.

**Review by VincenzoArceri — APPROVED** [review]

**Pull request merged**